Library/Homebrew/cmd/test-bot.rb

```diff
@@ -944,11 +953,21 @@ def test_bot
 
           if step.has_output?
             # Remove invalid XML CData characters from step output.
-            output = step.output.delete("\000\a\b\e\f\x2\x1f")
+            valid_xml_chars = ""
+            if ruby_has_encoding?
+              # This is the regex for valid XML chars, but only works in Ruby 2.0+
+              # /[]\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u{10000}-\u{10FFFF}]/
+              # For 1.9 compatibility, use the inverse of that, which staus under \u10000
+              invalid_xml_pat = /[\x00-\x08\x0B\x0C\x0E-\x1F\uD800-\uDFFF\uFFFE\uFFFF]/
+              output = step.output.gsub(invalid_xml_pat, "\uFFFD")
+            else
+              output = step.output.delete("\000\a\b\e\f\x2\x1f")
+            end
 
             if output.bytesize > BYTES_IN_1_MEGABYTE
-              output = "truncated output to 1MB:\n" \
-                + output.slice(-BYTES_IN_1_MEGABYTE, BYTES_IN_1_MEGABYTE)
+              slice_start = [0, -BYTES_IN_1_MEGABYTE].max
```

Comment:
Isn't that always going to return 0? Shouldn't this be `[0, output.length - BYTES_IN_1_MEGABYTE].max`? Does that already account for the case that for a typical UTF-8 string `bytesize > length` and does it make sure not to cut into the middle of some UTF-8 byte sequence or a single code unit?

And more Unicode complications: What if that cuts into the middle of some grapheme cluster? It won't make the UTF-8 invalid, but it will disturb the output at the cut-off point. (Probably details we shouldn't care about.)